### PR TITLE
Beam search seq2seq from torch agent

### PR DESCRIPTION
**Pull request by uralik**
Beam search supporting min length, max length, n best lists, logging as dot format image, works with mini batches

### CODE REVIEW

File: parlai/core/torch_agent.py [hunk 1]
```diff
@@ -864,6 +864,11 @@ def get_backtrack_from_current_step(self):
 
     def advance(self, softmax_probs):
         voc_size = softmax_probs.size(-1)
+        current_length = len(self.all_scores) - 1
+        if current_length < self.min_length:
+            # penalize all eos probs to make it decode longer
+            for hyp_id in range(softmax_probs.size(0)):
+                softmax_probs[hyp_id][self.eos] = -1e20
```
[the remark this answers is not in the document]
Comment (alexholdenmiller) [reply parlai/core/torch_agent.py]: ha I added another reference to -1e20 to my diff too.

File: parlai/agents/seq2seq/seq2seq.py [hunk 2]
```diff
@@ -97,9 +98,15 @@ def add_cmdline_args(cls, argparser):
                                 'softmax (see arxiv.org/abs/1711.03953).')
         agent.add_argument('--beam-size', type=int, default=1,
                            help='Beam size, if 1 then greedy search')
-        agent.add_argument('--beam-log-freq', type=float, default=0.0,
+        agent.add_argument('--beam-dot-log', type='bool', default=False,
                            help='The portion of beams to dump from minibatch '
-                                'into model_name.beam_dump folder')
+                                'into /tmp folder')
```
Comment (alexholdenmiller) [inline parlai/agents/seq2seq/seq2seq.py]: should rewrite text if this is a bool not a portion

File: parlai/agents/seq2seq/seq2seq.py [hunk 3]
```diff
@@ -97,9 +98,15 @@ def add_cmdline_args(cls, argparser):
...
+        agent.add_argument('--beam-min-n-best', type=int, default=3,
+                           help='Minimum number of nbest candidates to achieve'
+                                'during the beam search')
```
Comment (alexholdenmiller) [inline parlai/agents/seq2seq/seq2seq.py]: 'to achieveduring the' (space missing)

File: parlai/agents/seq2seq/seq2seq.py [hunk 4]
```diff
@@ -97,9 +98,15 @@ def add_cmdline_args(cls, argparser):
...
+        agent.add_argument('--beam-min-length', type=int, default=3,
+                           help='Minimum length of prediction to be generated by'
+                                'the beam search')
```
Comment (alexholdenmiller) [inline parlai/agents/seq2seq/seq2seq.py]: same here

File: parlai/agents/seq2seq/seq2seq.py [hunk 5]
```diff
@@ -143,19 +150,31 @@ def __init__(self, opt, shared=None):
...
             # set up shared properties
             self.model = shared['model']
             self.metrics = shared['metrics']
             states = shared.get('states', {})
+            if self.beam_dot_log is True:
+                self.beam_dot_dir = shared['beam_dot_dir']
 
         else:
             # this is not a shared instance of this class, so do full init
+            if self.beam_dot_log is True:
```
Comment (alexholdenmiller) [inline parlai/agents/seq2seq/seq2seq.py]: `if self.beam_dot_log:`

**Comment by uralik** [comment]
okay fixed wrt your comments, and should pass peps tests now

File: parlai/agents/seq2seq/seq2seq.py [hunk 6]
```diff
@@ -375,13 +397,150 @@ def _pick_cands(self, cand_preds, cand_inds, cands):
...
+        cand_scores = None
         cand_params = self._build_cands(batch)
-        out = self.model(batch.text_vec, ys=None, cand_params=cand_params)
-        scores, cand_scores = out[0], out[1]
-        _, preds = scores.max(2)
+        if self.beam_size == 1:
+            out = self.model(batch.text_vec, ys=None, cand_params=cand_params)
+            scores, cand_scores = out[0], out[1]
+            _, preds = scores.max(2)
+            preds = preds.cpu()
```
Comment (alexholdenmiller) [inline parlai/agents/seq2seq/seq2seq.py]: did you mean to call the `greedy_search` function here?

Comment (alexholdenmiller) [reply parlai/agents/seq2seq/seq2seq.py]: also don't need to call preds.cpu()

**Pull request merged**